cpp/src/arrow/filesystem/filesystem.h

```diff
@@ -387,6 +387,14 @@ Result<std::shared_ptr<FileSystem>> FileSystemFromUriOrPath(
 
 /// @}
 
+/// \brief Copy files from one FileSystem to another
+ARROW_EXPORT
+Status CopyFiles(const std::shared_ptr<FileSystem>& src_fs,
```

Comment:
We may introduce a `FileLocator` struct or something that would be a `filesystem, path` pair. Would also be useful with `FromURI`.

Reply:
Would you then want the signature to be `Status CopyFiles(const std::vector<FileLocator>& sources, const std::vector<FileLocator>& destinations, ...)`?

https://issues.apache.org/jira/browse/ARROW-9909

Reply:
That sounds ok, assuming having potentially different filesystems in those vectors is not a problem.

Reply:
I've wanted some kind of object to contain fs and path (like what you get from FileSystemFromUri) for a while too. Yesterday in fact I was experimenting with using SubTreeFileSystem to hold that, since it initializes with a fs and path. That mostly works as expected (for the operations I was trying at least).